File: app/models/booking.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship

from app.db.base import Base

class Booking(Base):
    __tablename__ = "bookings"

    id = Column(Integer, primary_key=True, index=True)
    slot_id = Column(Integer, ForeignKey("slots.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    booking_time = Column(DateTime, default=datetime.utcnow, nullable=False)
    status = Column(String, default="booked", nullable=False)

    # Relationships
    slot = relationship("Slot", back_populates="bookings")
    user = relationship("User", back_populates="bookings")

    # Unique constraint to prevent duplicate bookings
    __table_args__ = (
        UniqueConstraint('slot_id', 'user_id', name='uq_user_slot'),
    )

File: app/models/booking.js
const mongoose = require('mongoose');

const BookingSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  slotId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Slot',
    required: true
  },
  status: {
    type: String,
    enum: ['confirmed', 'cancelled', 'completed'],
    default: 'confirmed'
  },
  bookingDate: {
    type: Date,
    default: Date.now
  }
}, { timestamps: true });

// Compound index to prevent duplicate bookings
BookingSchema.index({ userId: 1, slotId: 1 }, { unique: true });

module.exports = mongoose.model('Booking', BookingSchema);
